### ISSUE TITLE
250812 Deep Learning 모델학습법_정서우

### ISSUE THREAD

**Issue by Seowoo-C** [AIBootcamp14/wiki-rider#26]
# 🧠 딥러닝 모델 학습법 

🗓 📅 학습 날짜  
2025-08-12  

📌 키워드 및 요약 정리  
**키워드**: Perceptron, Multi-Layer Perceptron(MLP), Activation Function, Gradient Descent, SGD, Backpropagation  

**요약**  
- 퍼셉트론: 선형 분류기, XOR 해결 불가  
- MLP: 은닉층 + 활성화 함수 → 비선형 패턴 학습 가능  
- 경사하강법: 손실 최소화를 위한 파라미터 업데이트  
- SGD: 데이터 일부로 빠르고 효율적인 학습  
- 역전파: 기울기 효율적 계산을 위한 체인 룰 기반 알고리즘  

### 📌 퍼셉트론
- **구조**: 입력 → 가중치  → 합산 + 편향  → 임계치 비교 → 스텝 함수로 0/1 출력  
- **선형 분리 가능**: AND, OR, NAND  
- **한계**: XOR 같은 **비선형 분리 문제**는 해결 불가  
- **기하학적 해석**: 2D에서 직선 하나로 영역 분리  

---

#### ▪ 정의 
- 입력층과 출력층 사이에 **하나 이상의 은닉층**을 둔 네트워크 
- 각 층의 뉴런이 이전 층 모든 뉴런과 연결되는 **계층형 신경망**  
- 실무에서는 **MLP ≈ Fully Connected Network**로 사용되는 경우가 많음  

#### ▪ 활성화 함수 필요성
- **비선형성 부여** → 복잡한 패턴 학습 가능  
- 활성화 함수가 없으면 다층이어도 선형식으로 환원되어 표현력 제한  

---

### 🔌 활성화 함수 예시 (MLP)

| 활성화 함수 | 출력 범위 | 주요 특징/목적 | 장점 | 단점 |
|-------------|----------|----------------|------|------|
| **Sigmoid** | (0, 1) | 연산 전체 **미분 가능**, 확률 해석 가능, **역전파 기법**과 연관 | 확률 표현 용이 | 포화 구간에서 기울기 소실, 출력 중심이 0이 아님 |
| **Tanh** | (-1, 1) | Sigmoid 변형, 출력 중심이 0에 가까움 | 학습 안정성↑, 중심 0 | 포화 구간에서 기울기 소실 |
| **ReLU** | [0, ∞) | 양수=그대로, 음수=0 | 계산 단순, 기울기 소실 완화 | 음수 입력 시 Dead Neuron 문제 |
| **Leaky ReLU** | (-∞, ∞) | ReLU 변형, 음수 영역에 작은 기울기 부여 | Dead Neuron 문제 완화 | 최적 기울기 α 선택 필요 |
| **ELU** | (-∞, ∞) | 음수 영역 완만한 지수 변화, 평균 0 근처 유지 | 학습 안정성↑ | 계산량 증가 |
| **Maxout** | (-∞, ∞) | 여러 선형 함수 중 최대값 출력 | 높은 표현력, 다양한 활성화 일반화 | 파라미터 수 증가, 계산량↑ |

### 📌 개념
- 경사하강법은 **기울기(gradient)가 감소하는 방향**으로 변수를 이동시켜 함수의 최소값을 찾는 최적화 알고리즘.
- 손실 함수(loss function)의 그래디언트가 감소하는 방향으로 **모델 파라미터(θ)**를 갱신하여 손실을 최소화.
- 손실함수 전체 모양이 부드러운 U자 모양임을 가정함. 
- 수식:
  
  $$
  \theta \leftarrow \theta - \eta \nabla L(\theta)
  $$

```
- θ : 모델 파라미터 (weights, biases)  
- η : 학습률(learning rate)  
- L(θ) : 손실 함수  
- ∇L(θ) : θ에 대한 손실 함수의 기울기   

```

### 📌 원리
1. 현재 위치에서 기울기를 계산
2. 기울기가 감소하는 방향(반대 방향)으로 파라미터 이동
3. 이동 크기는 학습률 η로 조절
4. 손실이 최소값에 가까워질 때까지 반복

#### 1. Local Minimum 문제
- **설명**: 시작 위치(초기 파라미터)에 따라 전역 최소값(Global Minimum)이 아닌 국소 최소값(Local Minimum)에 수렴 가능

- **해결방안**:
  - 파라미터 초기화를 잘 수행 (Xavier, He 초기화 등 적용)
  - Pretraining 활용 가능
  - 모델 구조 변경 → 네트워크 아키텍처 변경 시 손실 함수 지형이 바뀌어 더 나은 최소값 도달 가능
  - Learning Rate 조정 (Learning Rate Decay, Warm-up, Scheduler 활용)

---

#### 2. Plateau(기울기 거의 0) 구간 문제
- **설명**: 평평한 영역(Plateau)에서 기울기가 거의 0이 되어 학습이 정체됨

- **해결방안**:
  - Momentum, Adam 등의 옵티마이저 활용 → 기울기가 작은 구간도 통과 가능
  - 학습률 변동(learning rate scheduling)으로 정체 구간 극복

---

#### 3. Learning Rate 선택 문제
- **설명**:
  - 너무 낮으면 → 수렴 속도 느림
  - 너무 높으면 → 발산하거나 최소값을 지나침

- **해결방안**:
  - 적절한 학습률 탐색(Learning rate finder)
  - Adaptive Learning Rate 기법(AdaGrad, RMSProp, Adam 등) 적용

---

### 📌 개념
SGD는 모든 데이터를 사용하지 않고 하나의 데이터 샘플 또는 **일부 데이터(m개)**를 사용하여 기울기를 계산하고,  
그때마다 파라미터를 업데이트하는 방식이다.  
이때 사용하는 데이터 묶음을 **배치(batch)** 또는 **미니배치(mini-batch)**라고 한다.  

- \( m = 1 \)이면 **온라인 학습(online/incremental learning)**이라고 부른다.

---

### 📌 특징
- 손실 함수를 1/m로 스케일링하여 계산
- 배치마다 다른 데이터 샘플이 선택되어 손실 함수 모양이 조금씩 변함
- 이로 인해 local minima에 빠질 가능성이 줄고 전역 최소값(Global Minimum)에 도달할 가능성이 커짐
- 메모리와 연산 효율성이 높음

---

### 🔄 파이프라인
1. 전체 데이터에서 **m개 샘플 랜덤 선택**
2. 해당 배치로 **손실 함수와 기울기 계산**
3. **파라미터 업데이트**
4. 배치를 바꿔가며 반복 → 1 epoch 완성

---

### 📊 GD와 비교

| 구분                | GD(Gradient Descent) | SGD(Stochastic Gradient Descent) |
|---------------------|----------------------|-----------------------------------|
| **데이터 사용**     | 전체 데이터          | 일부 데이터(배치)                 |
| **업데이트 횟수**   | 1 epoch당 1회        | 1 epoch당 여러 회                  |
| **메모리 사용**     | 큼                   | 적음                              |
| **수렴 속도**       | 안정적이나 느림       | 빠르지만 진동 가능성               |
| **Local minima 탈출** | 어려움               | 가능성 높음                        |

---

### 📌 개요
역전파는 신경망의 출력에서 입력 방향으로 오차를 전파하며, 가중치와 편향에 대한 **손실 함수의 기울기(gradient)**를 계산하는 알고리즘.

다층 퍼셉트론(MLP) 구조에서 주로 사용되며, **체인 룰(Chain Rule)**을 기반으로 각 층의 파라미터를 효율적으로 업데이트.

---

#### 1. 샘플별 손실 합 가정
전체 손실은 각 샘플 손실의 합으로 표현 가능 → 전체 기울기 대신 개별 샘플 기울기만 계산해도 됨.

#### 2. 출력층 의존성 가정
각 샘플의 손실은 최종 출력층의 활성화 값 \( a^L \)에 대한 함수로 정의.  
중간층 출력에 대해서는 정의하지 않음.

---
### 📌 역전파의 기본 방정식 
추가로 참고하시면 좋을 것 같아 링크를 첨부합니다. 
https://wikidocs.net/37406 

### 📌 인사이트 
- **Low-activation neurons**: 활성화가 거의 0인 뉴런은 가중치 업데이트 영향이 거의 없음.
- **Saturated neurons**: 시그모이드 출력이 0 또는 1 근처이면 미분 값이 거의 0이 되어 학습이 느려짐.

---

❓ 어려웠던 부분  

- 전체적으로 새로운 개념이 계속 추가되다보니 낯설고 어렵게 느껴졌습니다.  
- 특히, 역전파 방정식을 이해하는 데 어려움이 있었습니다.  
- 이론과 실습 사이의 간격이 커 실습을 이해하는 것이 쉽지 않았습니다.  

💬 질문 또는 토론 유도  

- **고정 학습률 vs Adaptive Learning Rate**  
  실무에서는 두 방식 중 어느 쪽을 더 자주 사용하는지, 그리고 프로젝트 특성이나 모델 구조에 따라 어떤 기준으로 선택하는지 궁금합니다.  

- **배치 크기와 학습률 조정**  
  GPU 메모리가 제한적인 상황에서 배치 크기를 줄여야 할 때, 학습률을 어떻게 조정하는 것이 안정적인 학습에 효과적인지 논의해보고 싶습니다.  

- **모델 복잡도와 과적합 방지**  
  은닉층 개수와 뉴런 수를 늘려 모델 복잡도를 높일 때, 과적합을 예방하기 위해 어떤 기법(정규화, 드롭아웃, 데이터 증강 등)을 우선 고려하는지 경험을 공유하고 싶습니다.

**Comment by parkjunohjjang12** [comment]
안녕하세요 정서우 님! 올려주신 글 잘 읽었습니다.

퍼셉트론부터 MLP, 활성화 함수, 경사하강법, 역전파까지 딥러닝 학습법의 주요 개념을 구조적으로 잘 정리해 주신 것 같습니다!
이번 피드백은 어려웠던 부분과 질문에 대한 의견, 그리고 학습 보완 팁 위주로 드리겠습니다.

---
## 1️⃣ 어려웠던 부분 피드백

역전파 방정식 이해에서 어려움을 느끼셨다고 하셨는데, 이는 대부분의 학습자들이 겪는 공통적인 난관입니다. 저 같은 경우에는 처음에는 backpropgation을 진행할 때 chain rule이 어떻게 동작하는지에 대한 어려움이 있었습니다. 이에 대한 피드백으로 앞서 민병호 님께 전달드린 내용과 같은 자료들을 전달 드리고자 하며, 이에 대해서 강의 자료들을 보면서 공부하시고 구현해보면 정말 많은 도움이 될 것이라고 생각합니다.

📄  [이준석 교수님 ML/DL Lec3: Neural Networks & Backpropagation](https://www.youtube.com/watch?v=O8FMi94Upxc&list=PL0E_1UqNACXAc-tb9z2069eDeIOpcO8A2&index=3
)
📄 [CS231n](https://cs231n.github.io/optimization-2/)  
📄 [Backpropagation Calculus – 3Blue1Brown](https://www.3blue1brown.com/lessons/backpropagation-calculus)  
📄 [NYU Deep Learning](https://atcold.github.io/NYU-DLSP21/)

해당 과정을 통해서 많은 도움을 받으시고 공부가 될 것이지만, 다른 방향으로도 한 번 공부하면 정서우 님께서 이해하시는데 더 도움이 될 것이라고 생각합니다! 이준석 교수님께서 올려주신 해당 강의 외에도 다른 기본적인 내용들도 많이 올려주셨으니 공부할 때 참고하시면 될 것 같습니다. 또한, NYU Deep Learning에서 backpropagation에 대한 구현 코드도 제공하였으니, 이에 대해 처음부터 scratch로 구현하면 더 수월하게 이해할 수 있을거라 생각하니 추천 드립니다!!

---
## 2️⃣ 질문 답변

**Q1. 고정 학습률 vs Adaptive Learning Rate**  
- 실무에서는 어떠한 방향으로 학습하냐에 따라 선택하는 것이 다르지만 저 같은 경우에는 AdamW나 LARS와 같은 Adaptive Learning Rate를 많이 사용하는 편입니다. 단 고정 학습률을 사용할 경우에는 Cosine Annealing이나 Warm-Up과 같은 추가적인 LR Scheular를 붙여 Vision Transformer와 같은 대규모 학습에서 자주 사용됩니다.

- 이에 대한 선택 기준은 데이터 규모, 모델 복잡도 등 학습 안전성 요구에 따라 결정된다고 보시면 될 것 같습니다! (Adaptive → 빠른 초기 수렴 / 고정+스케줄러 → 안정적 수렴)  

- 추후에 모델 학습을 진행할 때 다양한 Optimizer를 가지고 학습하면서 해당 Task에 어떠한 방법이 가장 효과적인지 판단하면서 진행하는 것을 추천 드립니다!

**Q2. 배치 크기와 학습률 조정** 
- 일반적으로 배치 크기를 절반으로 키우면 학습률을 절반으로 늘리는 것이 안정적입니다. 
- 예: 기존 batch size / learning rate가 `256 / 0.01` → 배치 크기 `512`로 변경 시 lr `0.02`로 조정.

**Q3. 모델 복잡도와 과적합 방지**  
- 은닉층과 뉴런수를 증강하는 parameter들을 늘릴 때, 저 같은 경우에는 Dropout을 먼저 고려하는 것이 좋을 것 같으며, 데이터 증강을 가장 후순위로 둘 것 같습니다! 데이터 증강의 경우 다양성을 키우기 위해 사용되는 방법으로, 일반적으로 은닉층과 뉴런수보다는 데이터 수를 고려하여 적용하는 것으로 보시면 좋을 것 같습니다.
-  다만 증강을 할 때 어떻게 증강을 할지에 대해서 생각하는 것도 중요하다고 생각됩니다. 너무 간단한 증강을 하거나 증강 없이 하는 것도 성능에 좋은 영향을 주기 힘들지만, 반대로 증강 강도가 너무 높아지면 오히려 학습이 어려워질 수도 있습니다. 증강에는 절대적인 정답이 없으므로, 다양한 실험을 통해 노하우를 쌓아가는 것이 중요하다고 생각합니다!

---

딥러닝을 처음 공부하시다 보면 어렵고 막막하게 느껴지실 거라 생각합니다. 저도 처음엔 이해되지 않는 개념들이 많았고, 이 모든 걸 다 알아야 하는건가 하는 고민이 많았습니다.

하지만 돌이켜보니 그런 어려움을 겪는 과정 자체가 저에게 많은 도움이 되었습니다. 모든 개념을 한 번에 완벽히 이해하려 하기보다는, 차근차근 하나씩 익혀가는 것이 중요하다고 생각합니다! 포기하지 마시고 꾸준히 학습하시면 분명 좋은 결과가 있을 거라 믿습니다! 

화이팅하세요! 💪

**Comment by Seowoo-C** [comment]
세심한 피드백 정말 감사드립니다!!
